File: ios/Sources/Application/EvidencePackFocus.swift
import Foundation
import OrynodeDomain

/// Query-aware evidence packaging: prefer hits that literally overlap the question,
/// and window excerpts around that overlap. Generic lexical focus — not entity correction.
///
/// Contract: focusing must not starve the model. A query hit on a short heading line
/// still expands to a budgeted context window inside the chunk (AskKB.Trace showed
/// page-43 evidence collapsed to `编辑反向代理服务器配置文件：` / 5 tokens — unusable).
public enum EvidencePackFocus {
    /// Soft floor used when scoring which hit is "richer" around the query.
    public static let scoringWindowCharacters = 160

    /// Stable re-rank: richer query-overlapping windows first, then retrieval score.
    public static func prioritize(
        _ hits: [KnowledgeSearchHit],
        query: String
    ) -> [KnowledgeSearchHit] {
        let needles = queryNeedles(query)
        guard !needles.isEmpty else { return hits }
        return hits.enumerated()
            .sorted { lhs, rhs in
                let left = overlapScore(lhs.element.chunk.text, needles: needles)
                let right = overlapScore(rhs.element.chunk.text, needles: needles)
                if left != right { return left > right }
                if lhs.element.score != rhs.element.score {
                    return lhs.element.score > rhs.element.score
                }
                return lhs.offset < rhs.offset
            }
            .map(\.element)
    }

    /// Center a budgeted window on the best query overlap; fall back to prefix.
    /// Never returns a lone short heading line when the chunk has more text.
    public static func excerpt(
        from text: String,
        query: String,
        maxCharacters: Int
    ) -> String {
        let budget = max(32, maxCharacters)
        let normalized = text
            .replacingOccurrences(of: "\r\n", with: "\n")
            .replacingOccurrences(of: "\r", with: "\n")
        guard !normalized.isEmpty else { return "" }

        let needles = queryNeedles(query)
        for needle in needles {
            guard let match = normalized.range(of: needle) else { continue }
            return characterWindow(around: match, in: normalized, maxCharacters: budget)
        }

        if normalized.count <= budget {
            return normalized
        }
        return String(normalized.prefix(budget))
    }

    // MARK: - Needles

    static func queryNeedles(_ query: String) -> [String] {
        QueryFocus.terms(from: query)
    }

    /// Prefer hits whose local window around the query is substantive (not a thin heading).
    private static func overlapScore(_ text: String, needles: [String]) -> Int {
        var best = 0
        for needle in needles {
            guard let match = text.range(of: needle) else { continue }
            let window = characterWindow(
                around: match,
                in: text,
                maxCharacters: scoringWindowCharacters
            )
            var score = needle.count * 10 + min(window.count, scoringWindowCharacters)
            let line = lineContaining(match, in: text)
            if line.hasSuffix("：") || line.hasSuffix(":") {
                score -= 80
            }
            if line.count < 20 {
                score -= 40
            }
            best = max(best, score)
        }
        return best
    }

    private static func lineContaining(
        _ match: Range<String.Index>,
        in text: String
    ) -> String {
        let lineStart = text[..<match.lowerBound].lastIndex(of: "\n").map { text.index(after: $0) }
            ?? text.startIndex
        let lineEnd = text[match.upperBound...].firstIndex(of: "\n") ?? text.endIndex
        return String(text[lineStart..<lineEnd]).trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func characterWindow(
        around match: Range<String.Index>,
        in text: String,
        maxCharacters: Int
    ) -> String {
        let matchLen = text.distance(from: match.lowerBound, to: match.upperBound)
        // Bias pad forward so a heading hit still pulls following body into the pack.
        let backPad = max(0, (maxCharacters - matchLen) / 4)
        let forwardPad = max(0, maxCharacters - matchLen - backPad)
        var start = match.lowerBound
        var remaining = backPad
        while remaining > 0, start > text.startIndex {
            start = text.index(before: start)
            remaining -= 1
        }
        var end = match.upperBound
        remaining = forwardPad
        while remaining > 0, end < text.endIndex {
            end = text.index(after: end)
            remaining -= 1
        }
        if let lineStart = text[..<start].lastIndex(of: "\n") {
            let candidate = text.index(after: lineStart)
            if text.distance(from: candidate, to: end) <= maxCharacters + 48 {
                start = candidate
            }
        } else if start != text.startIndex, text.distance(from: text.startIndex, to: end) <= maxCharacters + 48 {
            start = text.startIndex
        }
        if let lineEnd = text[end...].firstIndex(of: "\n") {
            let candidateEnd = lineEnd
            if text.distance(from: start, to: candidateEnd) <= maxCharacters + 48 {
                end = candidateEnd
            }
        }
        var sliced = String(text[start..<end]).trimmingCharacters(in: .whitespacesAndNewlines)
        if sliced.count > maxCharacters {
            sliced = String(sliced.prefix(maxCharacters))
        }
        return sliced
    }

    private static func collapseWhitespace(_ text: String) -> String {
        text
            .replacingOccurrences(of: "\u{00A0}", with: " ")
            .split(whereSeparator: \.isWhitespace)
            .joined(separator: " ")
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }
}

File: android/application/src/main/kotlin/ai/orynode/mobile/application/EvidencePackFocus.kt
package ai.orynode.mobile.application

import ai.orynode.mobile.domain.KnowledgeSearchHit

/**
 * Query-aware evidence packaging: prefer hits that literally overlap the question,
 * and window excerpts around that overlap. Must not starve the model with a thin heading.
 */
object EvidencePackFocus {
    const val SCORING_WINDOW_CHARACTERS = 160

    private val definitionCueRegex = Regex(
        "是指|指的是|定义为|又称|全称|即是|也就是|是一[种个类]|retrieval[- ]augmented",
        RegexOption.IGNORE_CASE,
    )

    fun prioritize(hits: List<KnowledgeSearchHit>, query: String): List<KnowledgeSearchHit> {
        val needles = QueryFocus.terms(from = query)
        if (needles.isEmpty()) return hits
        return hits.withIndex()
            .sortedWith { lhs, rhs ->
                val left = overlapScore(lhs.value.chunk.text, needles)
                val right = overlapScore(rhs.value.chunk.text, needles)
                when {
                    left != right -> right.compareTo(left)
                    lhs.value.score != rhs.value.score -> rhs.value.score.compareTo(lhs.value.score)
                    else -> lhs.index.compareTo(rhs.index)
                }
            }
            .map { it.value }
    }

    fun excerpt(from: String, query: String, maxCharacters: Int): String {
        val budget = maxOf(32, maxCharacters)
        val normalized = from.replace("\r\n", "\n").replace("\r", "\n")
        if (normalized.isEmpty()) return ""
        val needles = QueryFocus.terms(from = query)
        val best = bestMatch(normalized, needles)
        if (best != null) {
            return characterWindow(
                text = normalized,
                matchStart = best.start,
                matchEnd = best.end,
                maxCharacters = budget,
            )
        }
        return if (normalized.length <= budget) normalized else normalized.take(budget)
    }

    private fun overlapScore(text: String, needles: List<String>): Int =
        bestMatch(text, needles)?.score ?: 0

    private data class Match(val start: Int, val end: Int, val score: Int)

    private fun bestMatch(text: String, needles: List<String>): Match? {
        var best: Match? = null
        for (needle in needles) {
            var from = 0
            while (true) {
                val index = text.indexOf(needle, startIndex = from)
                if (index < 0) break
                val end = index + needle.length
                val window = characterWindow(
                    text = text,
                    matchStart = index,
                    matchEnd = end,
                    maxCharacters = SCORING_WINDOW_CHARACTERS,
                )
                var score = needle.length * 10 + minOf(window.length, SCORING_WINDOW_CHARACTERS)
                val line = lineContaining(text, index, end)
                if (line.endsWith("：") || line.endsWith(":")) score -= 80
                if (line.length < 20) score -= 40
                if (definitionCueRegex.containsMatchIn(window)) score += 60
                if (best == null || score > best.score) {
                    best = Match(index, end, score)
                }
                from = index + 1
            }
        }
        return best
    }

    private fun lineContaining(text: String, matchStart: Int, matchEnd: Int): String {
        val lineStart = text.lastIndexOf('\n', startIndex = (matchStart - 1).coerceAtLeast(0))
            .let { if (it < 0 || matchStart == 0) 0 else it + 1 }
        val lineEnd = text.indexOf('\n', startIndex = matchEnd).let { if (it < 0) text.length else it }
        return text.substring(lineStart, lineEnd).trim()
    }

    private fun characterWindow(
        text: String,
        matchStart: Int,
        matchEnd: Int,
        maxCharacters: Int,
    ): String {
        val matchLen = matchEnd - matchStart
        val backPad = maxOf(0, (maxCharacters - matchLen) / 4)
        val forwardPad = maxOf(0, maxCharacters - matchLen - backPad)
        var start = (matchStart - backPad).coerceAtLeast(0)
        var end = (matchEnd + forwardPad).coerceAtMost(text.length)
        val lineStart = text.lastIndexOf('\n', startIndex = (start - 1).coerceAtLeast(0))
        if (lineStart >= 0) {
            val candidate = lineStart + 1
            if (end - candidate <= maxCharacters + 48) start = candidate
        } else if (start != 0 && end <= maxCharacters + 48) {
            start = 0
        }
        val lineEnd = text.indexOf('\n', startIndex = end)
        if (lineEnd >= 0 && lineEnd - start <= maxCharacters + 48) {
            end = lineEnd
        }
        var sliced = text.substring(start, end).trim()
        if (sliced.length > maxCharacters) sliced = sliced.take(maxCharacters)
        return sliced
    }
}
